src/Language/DifferentialDatalog/CompileJava.hs

```diff
@@ -0,0 +1,487 @@
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+-}
+
+{-# LANGUAGE RecordWildCards, FlexibleContexts, LambdaCase, TupleSections, OverloadedStrings, TemplateHaskell, QuasiQuotes #-}
+
+{- |
+Module     : CompileJava
+Description: Compile 'DatalogProgram' to Java API.
+-}
+
+module Language.DifferentialDatalog.CompileJava (
+    compileJava,
+) where
+
+import Prelude hiding((<>))
+import Control.Monad.State
+import Text.PrettyPrint
+import Data.Tuple
+import Data.Tuple.Select
+import Data.Maybe
+import Data.List
+import Data.Int
+import Data.Word
+import Data.Bits hiding (isSigned)
+import Data.List.Split
+import Data.FileEmbed
+import Data.String.Utils
+import Data.Char
+import System.FilePath
+import System.Directory
+import qualified Data.ByteString.Char8 as BS
+import Numeric
+import qualified Data.Set as S
+import qualified Data.Map as M
+import qualified Data.Graph.Inductive as G
+import qualified Data.Graph.Inductive.Query.DFS as G
+import Debug.Trace
+import Text.RawString.QQ
+import Data.WideWord
+import Debug.Trace
+
+import Language.DifferentialDatalog.PP
+import Language.DifferentialDatalog.Name
+import Language.DifferentialDatalog.Pos
+import Language.DifferentialDatalog.Ops
+import Language.DifferentialDatalog.Util
+import Language.DifferentialDatalog.Syntax
+import Language.DifferentialDatalog.Parse
+import Language.DifferentialDatalog.NS
+import Language.DifferentialDatalog.Expr
+import Language.DifferentialDatalog.DatalogProgram
+import Language.DifferentialDatalog.Relation
+import Language.DifferentialDatalog.Optimize
+import Language.DifferentialDatalog.Module
+import Language.DifferentialDatalog.ECtx
+import Language.DifferentialDatalog.Type
+import Language.DifferentialDatalog.Rule
+
+-- generate a Java program from a DDlog program
+-- The Java program has some utility function to more easily create
+-- records for the DDlog relations.  Note that Java does not support
+-- all DDlog constructs
+compileJava :: DatalogProgram -> String -> IO ()
+compileJava d sourceName = do
+    let content = generateJava d sourceName in
+         writeFile (capitalize sourceName ++ ".java") (render content)
+
+-- Generate a Java document from a program given the source filename
+generateJava :: DatalogProgram -> String -> Doc
+generateJava d sourceName =
+    preamble sourceName $+$
+    (braces' $ body d)
+
+-- Generates the preamble of a Java file
+preamble :: String -> Doc
+preamble sourceName =
+    "import java.util.*;"                       $+$
+    "import ddlogapi.DDlogRecord;\n"            $+$
```

Comment:
No need for `\n` here and in the next line.

Reply:
I wanted an empty line